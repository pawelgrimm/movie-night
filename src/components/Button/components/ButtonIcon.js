import React from "react";

const ButtonIcon = ({ children, ...rest }) => {
  return <>{children}</>;
};

export default ButtonIcon;
