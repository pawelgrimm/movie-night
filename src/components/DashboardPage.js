import React from "react";
import MovieSearch from "./MovieSearch";
import MovieList from "./MovieList";

export const DashboardPage = () => {
  return (
    <div className="content-container">
      <MovieSearch numResults={5} />
      <MovieList />
    </div>
  );
};

export default DashboardPage;
