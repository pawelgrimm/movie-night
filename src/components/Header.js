import React from "react";
import { Link } from "react-router-dom";

export const Header = () => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link
          className="header__title"
          to="/dashboard"
        >
          <h1>Movie Night</h1>
        </Link>
      </div>
    </div>
  </header>
);

export default Header;
