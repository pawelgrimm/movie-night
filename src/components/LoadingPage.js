import React from "react";
import Loader from "./Loader";

const LoadingPage = () => (
  <div className="loader">
    <Loader />
  </div>
);

export default LoadingPage;
