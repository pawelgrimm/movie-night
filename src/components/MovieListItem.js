import React, { useState } from "react";
import { connect } from "react-redux";
import { removeMovie } from "../actions/movie";
import ReactImageFallback from "react-image-fallback";

const mapDispatchToProps = (dispatch) => ({
  removeMovie: (id) => dispatch(removeMovie(id)),
});

export const MovieListItem = ({ movie, removeMovie }) => {
  const [selected, setSelected] = useState(false);

  const onButtonClick = () => {
    removeMovie(movie.id);
  };
  const onItemClick = () => {
    setSelected(!selected);
  };

  return (
    <div
      className="list-item"
      onClick={onItemClick}
      style={{
        background:
          selected &&
          `url(https://image.tmdb.org/t/p/w500${movie.backdrop_path})`,
      }}
    >
      {selected ? (
        <>
          <ReactImageFallback
            src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`}
            fallbackImage="/images/default_poster.png"
            initialImage="/images/loader.gif"
            alt={`poster image for ${movie.title}`}
            className="search-result-item__poster"
          />
          <div className="search-result-item__info">
            <h3>{movie.title}</h3>
            <span>{movie.release_date}</span>
            <p>{movie.overview}</p>
            <p>{movie.t}</p>
          </div>
          <div className="search-result-item__button">
            <button
              className="button button--secondary"
              onClick={onButtonClick}
            >
              -
            </button>
          </div>
        </>
      ) : (
        <>
          <ReactImageFallback
            src={`https://image.tmdb.org/t/p/w92${movie.poster_path}`}
            fallbackImage="/images/default_poster.png"
            initialImage="/images/loader.gif"
            alt={`poster image for ${movie.title}`}
            className="search-result-item__poster"
          />
          <div className="search-result-item__info">
            <h3>{movie.title}</h3>
            <span>{movie.release_date}</span>
            <p>{movie.overview}</p>
          </div>
          <div className="search-result-item__button">
            <button
              className="button button--secondary"
              onClick={onButtonClick}
            >
              -
            </button>
          </div>
        </>
      )}
    </div>
  );
};

export default connect(undefined, mapDispatchToProps)(MovieListItem);
