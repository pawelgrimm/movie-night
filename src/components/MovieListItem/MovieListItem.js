import React, { useState } from "react";
import { connect } from "react-redux";
import OutsideClickHandler from "react-outside-click-handler";
import ItemCollapsed from "./MovieListItemCollapsed";
import ItemExpanded from "./MovieListItemExpanded";
import Loader from "../Loader/Loader";

const mapStateToProps = ({ videoModal }) => ({
  isModalOpen: videoModal.isOpen,
});

export const MovieListItem = ({
  isLoading,
  movie,
  isModalOpen,
  collapsedButtons,
  expandedButtons,
}) => {
  if (isLoading) {
    return <Loader />;
  }

  const [isExpanded, setExpanded] = useState(false);

  // Collapsed
  const expandItem = () => setExpanded(true);
  const CollapsedComponent = (
    <ItemCollapsed
      title={movie.title}
      rating={movie.rating}
      releaseDate={movie.release_date}
      genres={movie.genres}
      runtime={movie.runtime}
      voteAverage={movie.vote_average}
      posterPath={movie.poster_path}
      onClick={expandItem}
      buttons={collapsedButtons}
    />
  );

  // Expanded
  const collapseItem = (event) => {
    // Don't collapse the item due to a child button click
    if (event.target.tagName === "BUTTON") return;
    setExpanded(false);
  };
  const ExpandedComponent = (
    <ItemExpanded
      movie={movie}
      onClick={collapseItem}
      buttons={expandedButtons}
    />
  );

  return (
    <OutsideClickHandler onOutsideClick={collapseItem} disabled={isModalOpen}>
      {isExpanded ? ExpandedComponent : CollapsedComponent}
    </OutsideClickHandler>
  );
};

export default connect(mapStateToProps)(MovieListItem);
