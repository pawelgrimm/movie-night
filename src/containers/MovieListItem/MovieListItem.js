import React, {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import OutsideClickHandler from "react-outside-click-handler";
import { fetchMovie as fetchMovieAction } from "../../services/movie/actions";
import Loader from "../../components/Loader/Loader";
import { useDispatch, useSelector } from "react-redux";
import { useModal } from "../Modal/ModalContext";
import ExpandedItem from "./components/MovieListItemExpanded";
import CollapsedItem from "./components/MovieListItemCollapsed";

const useMovie = (id, setMovie) => {
  const dispatch = useDispatch();
  const fetchMovie = useCallback(
    (id) => {
      dispatch(fetchMovieAction(id));
    },
    [dispatch, id]
  );
  useEffect(() => {
    fetchMovie(id);
  }, []);

  const movieObject = useSelector(({ movies }) => movies[id]);

  useEffect(() => {
    setMovie(movieObject);
  }, [movieObject]);
};

const useExpandable = () => {
  const [isExpanded, setExpanded] = useState(false);
  return {
    isExpanded,
    setExpanded: () => setExpanded(true),
    setCollapsed: () => setExpanded(false),
  };
};

const MovieListItem = ({
  id,
  renderExpandedItems = () => {},
  renderCollapsedItems = () => {},
}) => {
  const { isExpanded, setExpanded, setCollapsed } = useExpandable();
  const [movie, setMovie] = useState(undefined);
  const { isOpen: isModalOpen } = useModal();

  useMovie(id, setMovie);

  return useMemo(
    () => (
      <li className="movie-list__item">
        {!movie ? (
          <Loader />
        ) : (
          <OutsideClickHandler
            onOutsideClick={setCollapsed}
            disabled={isModalOpen}
          >
            {isExpanded ? (
              <ExpandedItem
                id={id}
                movie={movie.info}
                renderAdditionalItems={renderExpandedItems}
                onClick={setCollapsed}
              />
            ) : (
              <CollapsedItem
                id={id}
                movie={movie.info}
                renderAdditionalItems={renderCollapsedItems}
                onClick={setExpanded}
              />
            )}
          </OutsideClickHandler>
        )}
      </li>
    ),
    [movie, isExpanded, isModalOpen]
  );
};

export default MovieListItem;
