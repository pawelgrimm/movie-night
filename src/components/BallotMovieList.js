import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import MovieListItem from "./MovieListItem";
import { ThumbsUp, ThumbsDown } from "react-feather";
import { openModal } from "../actions/videoModal";
import { getMoviesOnBallot } from "../firebase/api";
import Loader from "./Loader";
import { hydrateMovies } from "../selectors/movie";
import { fetchMovie } from "../actions/movie";

const mapStateToProps = ({ movies }) => ({
  savedMovies: movies,
});

const mapDispatchToProps = (dispatch) => ({
  openModal: (videos) => dispatch(openModal(videos)),
  fetchMovie: (id) => dispatch(fetchMovie(id)),
});

export const BallotMovieList = ({ savedMovies, openModal, fetchMovie }) => {
  const [loaded, setLoaded] = useState(false);
  const [movies, setMovies] = useState([]);
  const [movieIds, setMovieIds] = useState([]);
  const { id } = useParams();

  useEffect(() => {
    getMoviesOnBallot(id).then((ref) => {
      const movieIds = ref.val();
      movieIds.forEach((id) => fetchMovie(id));
      setMovieIds(movieIds);
      setMovies(hydrateMovies(savedMovies, movieIds));
      setLoaded(true);
    });
  }, []);

  useEffect(() => {
    if (loaded) {
      setMovies(hydrateMovies(savedMovies, movieIds));
    }
  }, [savedMovies]);

  if (!loaded) {
    return <Loader />;
  }

  return (
    <div className="list-body">
      {movies.map((movie) => {
        const id = movie.info.id;
        return (
          <MovieListItem
            key={id}
            isLoading={!movie.isSaved}
            movie={movie.info}
            collapsedButtons={collapsedButtons(id)}
            expandedButtons={expandedButtons(
              id,
              movie.info.trailers,
              openModal
            )}
          />
        );
      })}
    </div>
  );
};

const collapsedButtons = (movieId) => makeButtons(movieId, true);

const expandedButtons = (movieId, videos, openModal) => [
  videos && videos.length > 0 && (
    <button
      className="button"
      onClick={(event) => {
        event.stopPropagation();
        openModal(videos);
      }}
    >
      Watch Trailer
    </button>
  ),
  ...makeButtons(movieId, false),
];

const makeButtons = (movieId, collapsed = true) => [
  <button
    className="button"
    onClick={(event) => {
      event.stopPropagation();
      console.log("thumbs up clicked for", movieId);
    }}
  >
    <ThumbsUp />
  </button>,
  <button
    className="button"
    onClick={(event) => {
      event.stopPropagation();
      console.log("thumbs up clicked for", movieId);
    }}
  >
    <ThumbsDown />
  </button>,
];

export default connect(mapStateToProps, mapDispatchToProps)(BallotMovieList);
