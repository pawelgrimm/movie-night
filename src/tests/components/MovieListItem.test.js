import { expect } from "@jest/globals";
import { shallow } from "enzyme";
import React from "react";
import { MovieListItem } from "../../components/MovieListItem";
import { movies } from "../fixtures/fixtures";

test("should render MovieListItem correctly with movie", () => {
  const wrapper = shallow(
    <MovieListItem movie={movies[2]} removeMovie={() => {}} />
  );
  expect(wrapper).toMatchSnapshot();
});
