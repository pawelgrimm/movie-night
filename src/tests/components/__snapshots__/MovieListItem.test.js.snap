// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MovieListItem correctly with movie 1`] = `
<div
  className="list-item"
  onClick={[Function]}
  style={
    Object {
      "background": false,
    }
  }
>
  <ReactImageFallback
    alt="poster image for Before Sunset"
    className="search-result-item__poster"
    fallbackImage="/images/default_poster.png"
    initialImage="/images/loader.gif"
    src="https://image.tmdb.org/t/p/w92/gycdE1ARByGQcK4fYR2mgpU6OO.jpg"
  />
  <div
    className="search-result-item__info"
  >
    <h3>
      Before Sunset
    </h3>
    <span>
      2004-02-10
    </span>
    <p>
      Nine years ago two strangers met by chance and spent a night in Vienna that ended before sunrise. They are about to meet for the first time since. Now they have one afternoon to find out if they belong together.
    </p>
  </div>
  <div
    className="search-result-item__button"
  >
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      -
    </button>
  </div>
</div>
`;
